src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -191,7 +191,7 @@ public static class TransitiveNoWarnUtils
                         // Merge the node's package specific no warn to the one in the path.
                         var mergedPackageSpecificNoWarn = MergePackageSpecificNoWarn(pathPackageSpecificNoWarn, nodePackageSpecificNoWarn);
 
-                        AddDependenciesToQueue(nodeDependencies,
+                        AddDependenciesToQueue(nodeDependencies.Where(i => !seen.ContainsKey(i.Name)),
```

Comment:
What happens if a dependency that was seen before has a different version than the one, we are processing currently?

Reply:
Good point.. looks to me like this should be based on PR, which are version specific.
https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#suppressing-nuget-warnings

Reply:
I learnt from https://github.com/NuGet/Home/issues/5740#issuecomment-332941509 comment that `NoWarn` can be set at the project level as a MSBuild property. This means that `NoWarn` applies to all the package dependencies including transitive ones.

Reply:
> What happens if a package dependency that was seen before has a different version than the one, we are processing currently?

That can't happen (currently at least. With package shading things will get more complex).  The point of restore is to resolve the graph and select a single version for each package. Different projects could use different versions, but each project goes through this method independently.